PSPs/drafts/psp-34.md

```diff
@@ -0,0 +1,914 @@
+      "type": {
+        "displayName": [
+           "Id"
+        ],
+        "type": "Id"
+      }
+    },
+    {
+      "label": "data",
+      "type": {
+        "displayName": [
+          "[u8]"
+        ],
+        "type": "[u8]"
+      }
+    }
+  ],
+  "docs": [
+    "Transfer approved or owned token from `from`.",
+    "",
+    "On success a `Transfer` event is emitted.",
+    "",
+    "# Errors",
+    "",
+    "Returns `TokenNotExists` error if `id` does not exist.",
+    "",
+    "Returns `NotApproved` error if `from` doesn't have allowance for transferring.",
+    "",
+    "Returns `SafeTransferCheckFailed` error if `to` doesn't accept transfer."
+  ],
+  "mutates": true,
+  "label": "PSP34::transfer_from",
+  "payable": false,
+  "returnType": {
+    "displayName": [
+      "Result"
+    ],
+    "type": 1
+  },
+  "selector": "0x718fd38b"
+}
+```
+
+#### PSP34Receiver
+`PSP34Receiver` is an interface for any contract that wants to support safe transfers from a PSP-34 token smart contract to avoid unexpected tokens in the balance of contract.
+This method is called before a transfer to ensure the recipient of the tokens acknowledges the receipt.
+
+##### **before_received**(operator: AccountId, from: AccountId, id: Id, data: [u8]) ➔ Result<(), PSP34ReceiverError>
+Selector: `0xbb7df780` - first 4 bytes of `blake2b_256("PSP34Receiver::before_received")`
+```json
+{
+  "args": [
+    {
+      "label": "operator",
+      "type": {
+        "displayName": [
+          "AccountId"
+        ],
+        "type": "AccountId"
+      }
+    },
+    {
+      "label": "from",
+      "type": {
+        "displayName": [
+          "AccountId"
+        ],
+        "type": "AccountId"
+      }
+    },
+    {
+      "label": "id",
+      "type": {
+        "displayName": [
+           "Id"
+        ],
+        "type": "Id"
+      }
+    },
+    {
+      "label": "data",
+      "type": {
+        "displayName": [
+          "[u8]"
+        ],
+        "type": "[u8]"
+      }
+    }
+  ],
+  "docs": [
+    "Ensures that the smart contract allows reception of PSP34 token(s).",
+    "Returns `Ok(())` if the contract allows the reception of the token(s) and Error `TransferRejected(String)` otherwise.",
+    "",
+    "This method will get called on every transfer to check whether the recipient in `transfer`",
+    "or `transfer_from` is a contract, and if it is, does it accept tokens.",
+    "This is done to prevent contracts from locking tokens forever.",
+    "",
+    "Returns `PSP34ReceiverError` if the contract does not accept the tokens."
+  ],
+  "mutates": true,
+  "label": "PSP34Receiver::before_received",
+  "payable": false,
+  "returnType": {
+    "displayName": [
+      "Result"
+    ],
+    "type": 2
+  },
+  "selector": "0xbb7df780"
+}
+```
+
+### Extension
+
+#### PSP34Metadata
+
+`PSP34Metadata` is a **recommended** extension for this Non-Fungible Token standard
+because it is main feature of the NFT.
+
+##### **total_supply**() ➔ Balance
```

Comment:
I think we should put `total_supply` into the main data interface. Convince me the opposite.

Reply:
Convince me "Why"=D
Maybe it is not a bad idea because `total_supply` easy can be implemented in the scope of the main interface. It is ++ during minting and -- during burning.
We moved that to `PSP34Metadata` because it is useless information for cross-contact communication(in the case of fungible tokens it is useful information because all tokens are equal. In the case of non-fungible it is not so useful) but it can be useful for UI(`Erc721` doesn't have `total_supply`) like attributes.

Reply:
The fact that it is easy to implement means that it belongs to the main interface. Not possible to put it into the metadata without excessive complexity added.